### ISSUE TITLE
Go to definition for TypeScript literal strings

### ISSUE THREAD

**Issue by KATT** [microsoft/TypeScript#49033]
> **$500 bounty** from my own pocket if it's done before August 2022 _(just putting a timeline so I won't be hunted about this two years from now, feel free to contact me ahead of working on it if you are curious if I'm still up for it!)_

## Feature request

Go to definition for TypeScript literal strings.

### Example

Given the below code, I would really like to be able to jump to the definition of the `'foo'`:

```ts
const obj = {
  foo: 'bar' as const,
  hello: 'world' as const,
}

function get<T, K extends keyof T>(obj: T, path: K): T[K] {
  return obj[path]
}

const val = get(obj, 'foo')
//                   👆👆👆 Is there a way of make jump to definition work here?
//                         I really want to CMD+click `'foo'` and jump to line #2
```

[**Link to TypeScript playground**](https://www.typescriptlang.org/play?ts=4.6.2#code/MYewdgzgLgBCBGArGBeGBvAsAKBjAZiCAFwwDk8AhgE5kyUQyiRQA0OeAFgKYA2vJcgHcQ1XgBM6DJuGjtsAXxw58AVzDAoAS3AwA5tygAeACqsYAaRjcAHlG5hxjANbcAniHwwTAPgAUCIikZjAADpRQnKQWAJTBANoWALoYHDDUhqrUYHBI8eGRSThK2DjM0DAAbpS8qPqGAUjmZIQgZDE4APSdeL19-f2AvBuAYjsjIzAAkoyR3Bn0MEKUbnBeALaUrjCIqquhMFAgMOLc+FpgWtq6ItTOMDwZAPxdPQOvb70T6dw1vMuLYLADjAAMIAWQAIgBqYC8LTAW4AAxaRDICPoji2Oz2QNhYG4MAAxAAmZTYcogXjcAB0Aj0fmqvA62G67xgAD0HkA).

### WebStorm Demo of the same code

In WebStorm, it actually works how I wish that VSCode worked, here's a video:

https://user-images.githubusercontent.com/459267/167470744-80bdd8af-b7fa-444a-84d5-7e3f83e7e003.mov

**Comment by KATT** [comment]
Cc @RyanCavanaugh is there a way to put bug bounties on issues in order to incentivize OSS-contributors to take on tasks ahead of the core team? I'd happily put forward a few hundred $ personally for this to be solved as it's a big pain point in my API design of [tRPC](https://trpc.io/)

**Comment by RyanCavanaugh** [comment]
Interesting question; we do not currently have a mechanism for that. Other projects have had people put up bounties on third-party sites, e.g. https://github.com/dotnet/runtime/issues/36748#issuecomment-1105290484 . I was actually just on an internal email thread about this and the general consensus was that we're OK with good-faith bounties being staked, but want to "stay neutral" in terms of e.g. deciding which of two contributors should be awarded a bounty if both put up a PR.

**Comment by KATT** [comment]
Alright, **I'll pay $500** if someone solves this. I will come back later when I can find some bug bounty site where I can get the money back if it's not resolved.

**Comment by TheoBr** [comment]
I'll throw in another $500 on this bounty @KATT

**Comment by Andarist** [comment]
I will give it a shot then 🤣

**Comment by JoshuaKGoldberg** [comment]
@Andarist and I chatted a bit about this one. We suspect it's not straightforward or trivial to implement, but hopefully doable in our time budgets.

## Prior Debugging Notes

We'd both separately done similar investigations. Take my rough notes with a grain of salt because I'm not on the TypeScript team and don't know anything for sure:

1. The _Go to Definition_ codefix comes from `goToDefinition.ts`
2. Eventually the call stack goes to `function getSymbol` -> `checker.getSymbolAtLocation`
    * In here, the `case SyntaxKind.StringLiteral` case falls through to `SyntaxKind.NumericLiteral`
    * Nothing within those two cases gets hit, so it returns `undefined`

### Backing context: 

* Literals don't have rich (per-node) symbol information. There's one global symbol per literal type.
* Individual nodes seem to store a grab-bag of information in `nodeLinks`, accessed via `getNodeLinks`.

## Potential Implementation Directions

We discussed these three, in increasing order of likelihood to succeed:

1. Create new symbols for the literals
1. See what type the parent location is
1. Node links / getNodeLinks

### Create new symbols for the literals

Is it possible that we could have more rich symbols per literal? Maybe yes, but that'd be a pretty big breaking change in how TypeScript and its consumers understand symbols. I'm under the impression many consumers -such as typescript-eslint- assume one symbol-per-literal, not one-symbol-per-literal-and-type. 👎 

### See what type the parent location is

Is there a compiler API for _"given a node, tell me what type it's expected to be"_? E.g. if you're looking at a function argument, determining the intended type of the parameter? Neither of us could find one.

If one exists: yes please, that would be an ideal solution!

If not: ah well.

### Augment node links

`NodeLinks` seems to already be a catch-all place to put information on nodes. Can we add something like a `resolvedLiteralType?: Type` property for literals?

`getNodeLinks` is called all over the place. Neither of us have investigated where adding this information for literals would be possible, if anywhere.

---

Again: don't take these notes too seriously; we might be totally off in our direction. But either way: progress! 🚀

**Comment by RyanCavanaugh** [comment]
> Is there a compiler API for "given a node, tell me what type it's expected to be"? 

This is `getContextualType`

**Comment by RyanCavanaugh** [comment]
I'm not even sure how this can be made to work in practice. Let's consider something like
```ts
const obj1 = {
  hello: 1
};
const obj2 = {
  foo: 0,
  hello: 1,
  baz: 2
};

declare let a: (keyof typeof obj1) & (keyof typeof obj2);
a = "hello"; // <-
```
Here, `a` has the literal type `"hello"`. It went through intersection reduction via an identity match between `"hello"` from `keyof typeof obj1` and `keyof typeof obj2`. You can't create new literal types for those keys depending on where they came from; this will have totally uncool performance characteristics (allocations, forcing intersection to do more work, and GC pressure). So you're necessarily left with just the bare literal type `"hello"` with no real information of where it came from.

**Comment by JoshuaKGoldberg** [comment]
So this is interesting. Yes, `checker.getContextualType` does retrieve the union type - thank you!
But the union type doesn't have a `.symbol` referring back to the keyof typeof.

In fact, literal values and operator-computed values such as `keyof typeof` don't have type definitions.
Trying to _Go to type definition_ on the following variables' names...

- `const valA = {};`: highlights the `{}` ✅
- `const valB = true;`: gives _No type definition found for 'valB'_ ❌
- `const valC = typeof valA;`: gives _No type definition found for 'valC'_ ❌

Is it intentional that the type doesn't have a symbol? Could we add one in as part of this change?

> `declare let a: (keyof typeof obj1) & (keyof typeof obj2);`

I would think the _Go to definition_ would be on the type annotation in full: `(keyof typeof obj1) & (keyof typeof obj2)`. That would be the `checker.getContextType` behavior, no?

**Comment by Andarist** [comment]
> I would think the Go to definition would be on the type annotation in full: (keyof typeof obj1) & (keyof typeof obj2). That would be the expected checker.getContextType behavior, no?

That matches my thinking too.

As to `getContextualType` - I was trying to play around with it before but I couldn't make any use of it. With the given simple `get` example if I call `typeChecker.getContextualType(node)` I just end up with the literal type itself. Should it be called on another object?

**Comment by RyanCavanaugh** [comment]
> Is it intentional that the type doesn't have a symbol?

Yes. Literal types and unions don't have symbols because there's not an originating declaration for them; both are interned.

**Comment by RyanCavanaugh** [comment]
https://en.wikipedia.org/wiki/Interning_(computer_science)

**Comment by sghsri** [comment]
Any updates on this, would really love to have this behavior!
